cpp/arcticdb/async/tasks.hpp

```diff
@@ -327,18 +327,18 @@ struct DecodeSlicesTask : BaseTask {
 };
 
 struct SegmentFunctionTask : BaseTask {
-    stream::StreamSource::ReadContinuation func_;
+    std::shared_ptr<stream::StreamSource::ReadContinuation> func_;
```

Comment:
There is a niche case that makes necessary the shared pointer, and not unique pointer. 

If we use a unique pointer instead, we return the future containing this code:

            [this, keys, &continuations](auto&& key_pos) {
                auto key = keys[key_pos];
                auto cont = std::move(continuations[key_pos]);
                return async::submit_io_task(async::ReadCompressedTask(key, library_, storage::ReadKeyOpts{})).via(&async::cpu_executor())
                .thenValue(SegmentFunctionTask{std::move(cont)});
            }, 

Until the task SegmentFunctionTask is not executed, there is a time where no unique pointer points to continuations, and therefore, this is destroyed, which is what we want to avoid